### PR TITLE
Failing transcripts should always show the failing stanza

### PR DESCRIPTION
**Pull request by pete-ts**
## Overview

Currently the `stack exec transcripts` command runs the content of `Transcripts.hs`.

1. find all `.md` files in the `transcripts folder`
2. build a "ucm transcript command for each"
3. run each command as a test and expect it to return exit code 0.
4. if changes to `ucm` have affected the output to one or more transcripts, you now need to commit these, or these are caught by the build system and a build failure is produced.

The above behaviour is maintained and not altered at all.

This PR aims to add a new section of tests in the `stack exec transcripts` project that will allow failing transcripts to run to completion while capturing the error messages produced and tracking them as output transcripts.

1. find all `.md` files in the `errors folder`
2. build a "ucm transcript command for each"
3. run each transcript as a test and only capture the standard error.
4. if changes to `ucm` have affected the stderr of the transcripts, commit as necessary, or get a build failure. 

Following that, I have managed to address aspects of #1295 as seen in the output files.
1. when a ucm or unison block fails the failing stanza is always displayed
2. the error message for unexpected success has been updated
3. the error message for a failing stanza has been updated

## Implementation notes

`Transcripts.hs`
[shellmet](https://hackage.haskell.org/package/shellmet-0.0.3.0/docs/Shellmet.html) does not allow for running a failing command and capturing the error, but achieving this is quite straightforward (as `shellmet` is a wrapper around `System.Process` functions).

`TranscriptParser`
1. the stanza being worked on is stored in an IORef
2. this makes it easy to append it during an error

## Interesting/controversial decisions

* This can probably be achieved by `bash` but I feel that the approach suggested fits in with the existing mechanism for testing transcripts. There is no need to maintain an additional testing structure in parallel, as the "error" scenarios are testing the transcript parser, under the existing `stack exec transcripts`

* The local `die` function had to be renamed for clarity as the name clashes with `System.Exit (die)`

## Test coverage

This PR adds a way to enable additional testing, as well as examples of what these tests would look like.

Finally the PR tests the fix for #1295 

## Loose ends
* Once these two are sorted, a change so that `Transcripts.runAndCaptureError` only succeeds if it gets a non zero exit code.

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/TranscriptParser.hs [hunk 1]
```diff
@@ -68,7 +69,7 @@ instance Show Stanza where
   show s = case s of
     Ucm _ _ cmds -> unlines [
       "```ucm",
-      show cmds,
+      foldl (\x y -> x ++ show y) "" cmds,
```
Comment (pete-ts) [inline parser-typechecker/src/Unison/Codebase/TranscriptParser.hs]: This means that when showing a complete failing stanza, there are no angle braces around each command

File: parser-typechecker/src/Unison/Codebase/TranscriptParser.hs [hunk 2]
```diff
@@ -150,10 +152,12 @@ run dir configFile stanzas codebase = do
       awaitInput = do
         cmd <- atomically (Q.tryDequeue cmdQueue)
         case cmd of
+          -- end of ucm block
           Just Nothing -> do
-            output "\n```\n" -- this ends the ucm block
-            writeIORef hidden Shown
+            output "\n```\n"
+            dieUnexpectedSuccess
```
Comment (pete-ts) [inline parser-typechecker/src/Unison/Codebase/TranscriptParser.hs]: At the end of the `ucm` block check if there was an unexpected success
There is no need to update the `hidden`, as the next stanza will do that.

File: parser-typechecker/src/Unison/Codebase/TranscriptParser.hs [hunk 3]
```diff
@@ -244,20 +242,45 @@ run dir configFile stanzas codebase = do
...
-      die = do
+      -- Looks at the current stanza and decides if it is contained in the
+      -- output so far. Appends it if not.
+      appendFailingStanza :: IO ()
+      appendFailingStanza = do
+        stanzaOpt <- readIORef mStanza
+        currentOut <- readIORef out
+        let stnz = maybe "" show (fmap fst stanzaOpt :: Maybe Stanza)
+        unless (stnz `isSubsequenceOf` concat currentOut) $
+          modifyIORef' out (\acc -> acc <> pure stnz)
```
Comment (pete-ts) [inline parser-typechecker/src/Unison/Codebase/TranscriptParser.hs]: This function is required to allow us to overcome the fact that some blocks are hidden, after the fact. When either of `dieWithMsg` or `dieUnexpectedSuccess` gets called, the `awaitInput` method has broken down a stanza, and is working on individual `ucm` commands in the stanza. 

The output so far is read, and if the stanza being worked on and containing an error is already printed (by an `outputEcho` earlier) it is not appended a second time.

File: unison-src/transcripts/errors/ucm-hide.output.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+and surface a helpful message.
+
+```ucm
+.> move.namespace foo bar
+```
+
+
+🛑
+
+The transcript failed due to an error encountered in the stanza above.
```
Comment (pete-ts) [inline unison-src/transcripts/errors/ucm-hide.output.md]: I believe this the expected behaviour for #1295 @aryairani

File: unison-src/transcripts/errors/unison-hide.output.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+```unison
+g 3
+```
+
+
+
+🛑
+
+The transcript failed due to an error encountered in the stanza above.
```
Comment (pete-ts) [inline unison-src/transcripts/errors/unison-hide.output.md]: unison:hide shows the stanza

File: unison-src/transcripts/errors/unison-hide-error.output.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+```unison
+myVal = 3
+```
+
+
+
+🛑
+
+The transcript failed to fail as unexpected in the stanza above.
```
Comment (pete-ts) [inline unison-src/transcripts/errors/unison-hide-error.output.md]: unison hide error shows the stanza and the new error message

File: unison-src/transcripts/errors/unison-hide-all-error.output.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+```unison
+myVal = 3
+```
+
+
+
+🛑
+
+The transcript failed to fail as unexpected in the stanza above.
```
Comment (pete-ts) [inline unison-src/transcripts/errors/unison-hide-all-error.output.md]: unison:hide:all:error shows the stanza and updated message

File: unison-src/transcripts/errors/ucm-hide-all-error.md [hunk 8]
```diff
@@ -8,5 +8,5 @@ then the transcript parser should print the stanza
 and surface a helpful message.
 
 ```ucm:hide:all:error
-.> ls
+.> history
```
Comment (pete-ts) [inline unison-src/transcripts/errors/ucm-hide-all-error.md]: @aryairani came back to this with a clear head, and managed to debug it. `ls` was returning an error, so both of these were actually interpreted as "an error occurs in a block where an error is expected."

**Comment by aryairani** [comment]
@pete-ts Sorry for the delay on this.  I will dig into it "soon".  😬

**Pull request merged**